### ISSUE TITLE
Latest mpv (12b90e) cannot render on integrated Intel-GPU anymore

### ISSUE THREAD

**Issue by aitte2** [mpv-player/mpv#5631]
### mpv version and platform

I built this commit: https://github.com/mpv-player/mpv/commit/12b90e744dd220a3386a8977138fba28708c8d21

With these options: `brew install mpv --HEAD --with-bundle --with-libaacs --with-libarchive --with-libbluray --with-libdvdnav --with-libdvdread --with-rubberband --with-uchardet --with-vapoursynth`.

When trying to run the final binary in macOS 10.12's terminal, I get "Illegal instruction: 4". Google says that it's Apple's beautifully unclear language for "the application tried to execute an instruction that is not supported by this OS version".

So I downgraded to 0.28.2 "Stable" instead, and built it with the exact same options, and it's running fine.

Ping @Akemi - perhaps this is due to the new Swift stuff I saw in recent commits? :\

**Comment by Akemi** [comment]
i can't help you if you don't even provide a crash report, backtrace and/or build log. i mean why do you even ignore most of the issue template?

**Comment by aitte2** [comment]
@Akemi There is no crash report. This is all that Apple has to say:

![asdf](https://user-images.githubusercontent.com/18485256/37308405-3af98926-263e-11e8-8728-0ca315f316ed.png)

I'll do a rebuild to provide a build log though.

**Comment by aitte2** [comment]
@Akemi Okay, here are the steps to reproduce:

1. Have macOS 10.12.6.

2. Run these commands:

```
git clone https://github.com/mpv-player/mpv.git
cd mpv
./bootstrap.py
./waf configure
./waf
cd build
./mpv
```

Result: `Illegal instruction: 4`

Then do the same with v0.28.2 "Stable". Runs perfectly.

If Google is right about "Illegal Instruction: 4", it means that a syscall has been added since 0.28.2, which is not supported on macOS 10.12.

Logs:

[buildlog.txt](https://github.com/mpv-player/mpv/files/1804530/buildlog.txt)

**Comment by Akemi** [comment]
what does `swift -version` say? try updating your command line tools, especially if it's a 3.x swift version.

you should keep in mind it works for me. i am even on the same OS version. so you reproduction steps are not enough. it completely depends on your setup.

**Comment by aitte2** [comment]
Hmm. Seems fine?

```
$ swift -version
Apple Swift version 4.0.3 (swiftlang-900.0.74.1 clang-900.0.39.2)
Target: x86_64-apple-macosx10.9
```

And I checked App Store for updates. There are no OS updates. Hmm...

**Comment by Akemi** [comment]
try this in your terminal and give me the output pls.
1. `lldb -- mpv`
2. `run`
3. `bt`

**Comment by Akemi** [comment]
also to make sure it isn't something in your config, try with `mpv --no-config`.

the swift build and linking flags would also be helpful. just deleted `build/osdep/macOS_swift.o` and `build/mpv`, rerun `./waf -v` and post me the output.

last, also try with try with `./waf configure --disable-macos-cocoa-cb` to test if it's a swift problem.

**Comment by aitte2** [comment]
Got back from supper! Thank you for all of the suggestions!

Alright, here's the results:

```
./mpv --no-config
Illegal instruction: 4
```

It was a good idea to check but sadly had nothing to do with the config. ;\

----

```
rm -f build/osdep/macOS_swift.o && rm -f build/mpv
./waf -v
```

Result: 
[buildswift.txt](https://github.com/mpv-player/mpv/files/1804653/buildswift.txt)

---

```
./waf configure --disable-macos-cocoa-cb
./waf
./build/mpv --version
```

Result:

```
mpv 0.28.0-374-g12b90e744d Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on Mon Mar 12 22:37:08 CET 2018
```

Yes, it's a swift issue. I am going to do the lldb thing now.

**Comment by aitte2** [comment]
Hmm... something about the video layer?

```
~/Downloads/mpv/build
+ $ lldb -- ./mpv
(lldb) target create "./mpv"
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/usr/local/Cellar/python@2/2.7.14_3/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 52, in <module>
    import weakref
  File "/usr/local/Cellar/python@2/2.7.14_3/Frameworks/Python.framework/Versions/2.7/lib/python2.7/weakref.py", line 14, in <module>
    from _weakref import (
ImportError: cannot import name _remove_dead_weakref
Current executable set to './mpv' (x86_64).
(lldb) run
Process 12950 launched: './mpv' (x86_64)
Process 12950 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x0000000100126ae4 mpv`function signature specialization <Arg[0] = Dead, Arg[1] = Dead> of macOS_swift.VideoLayer.copyCGLPixelFormat(forDisplayMask: Swift.UInt32) -> Swift.OpaquePointer + 1220
mpv`function signature specialization <Arg[0] = Dead, Arg[1] = Dead> of macOS_swift.VideoLayer.copyCGLPixelFormat(forDisplayMask: Swift.UInt32) -> Swift.OpaquePointer:
->  0x100126ae4 <+1220>: ud2    
    0x100126ae6 <+1222>: nopw   %cs:(%rax,%rax)

mpv`function signature specialization <Arg[0] = Dead> of macOS_swift.Window.keepAspect.didset : Swift.Bool:
    0x100126af0 <+0>:    pushq  %rbp
    0x100126af1 <+1>:    movq   %rsp, %rbp
Target 0: (mpv) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
  * frame #0: 0x0000000100126ae4 mpv`function signature specialization <Arg[0] = Dead, Arg[1] = Dead> of macOS_swift.VideoLayer.copyCGLPixelFormat(forDisplayMask: Swift.UInt32) -> Swift.OpaquePointer + 1220
    frame #1: 0x00000001000f0ae9 mpv`@objc macOS_swift.VideoLayer.copyCGLPixelFormat(forDisplayMask: Swift.UInt32) -> Swift.OpaquePointer + 9
    frame #2: 0x00007fff8730bf10 QuartzCore`CAOpenGLLayerDraw(CAOpenGLLayer*, double, CVTimeStamp const*, unsigned int) + 292
    frame #3: 0x00007fff8730bcd3 QuartzCore`-[CAOpenGLLayer _display] + 351
    frame #4: 0x00000001000f0c8d mpv`@objc macOS_swift.VideoLayer.display() -> () + 77
    frame #5: 0x00007fff8731f042 QuartzCore`CA::Layer::display_if_needed(CA::Transaction*) + 572
    frame #6: 0x00007fff8731f16d QuartzCore`CA::Layer::layout_and_display_if_needed(CA::Transaction*) + 35
    frame #7: 0x00007fff87314924 QuartzCore`CA::Context::commit_transaction(CA::Transaction*) + 280
    frame #8: 0x00007fff872097a1 QuartzCore`CA::Transaction::commit() + 475
    frame #9: 0x00007fff7f2488b1 AppKit`__37+[NSDisplayCycle currentDisplayCycle]_block_invoke.31 + 323
    frame #10: 0x00007fff8133fde7 CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
    frame #11: 0x00007fff8133fd57 CoreFoundation`__CFRunLoopDoObservers + 391
    frame #12: 0x00007fff81320b09 CoreFoundation`__CFRunLoopRun + 873
    frame #13: 0x00007fff81320544 CoreFoundation`CFRunLoopRunSpecific + 420
    frame #14: 0x00007fff8087febc HIToolbox`RunCurrentEventLoopInMode + 240
    frame #15: 0x00007fff8087fbf9 HIToolbox`ReceiveNextEventCommon + 184
    frame #16: 0x00007fff8087fb26 HIToolbox`_BlockUntilNextEventMatchingListInModeWithFilter + 71
    frame #17: 0x00007fff7ee16a54 AppKit`_DPSNextEvent + 1120
    frame #18: 0x00007fff7f5927ee AppKit`-[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 2796
    frame #19: 0x00007fff7ee0b3db AppKit`-[NSApplication run] + 926
    frame #20: 0x00000001000e351f mpv`cocoa_main [inlined] cocoa_run_runloop at macosx_application.m:261 [opt]
    frame #21: 0x00000001000e34ef mpv`cocoa_main(argc=<unavailable>, argv=<unavailable>) at macosx_application.m:358 [opt]
    frame #22: 0x00007fff96eea235 libdyld.dylib`start + 1
    frame #23: 0x00007fff96eea235 libdyld.dylib`start + 1
(lldb) 
```

**Comment by aitte2** [comment]
I have narrowed down the issue.

1. Use https://github.com/steveschow/gfxCardStatus/releases/tag/v2.4.2i (that exact version is a sierra-fixed fork of the popular gfxCardStatus tool). It lets you choose Integrated to run the entire OS on the Intel GPU, or Dedicated to run it all on the dedicated Geforce GPU in Macbook Pros. Or Dynamic (OS X default) to let the OS auto-switch).

2. Run it and set it to "Integrated Only". This is what I run with 24/7 to save battery. I never need the dedicated GPU.

3. Run the latest mpv. It crashes with "Illegal Instruction: 4".

4. Try "Dynamic Switching" or "Dedicated Only", and run mpv again. It will work.

This means: Since v0.28.2, mpv has gained a dependency on "cannot render to internal GPU anymore"... Hmm. Interesting. Any idea why that happened? The older versions render very well via the internal Intel GPU. Performance and battery life (and noise) was great, since temperatures and battery usage stayed low. Mpv ran fine with the vo_gpu renderer on that Intel GPU, and the only thing I missed out on was advanced shader fx.

I really hope the Swift settings can be relaxed so that it works on the Intel GPU too again. The dedicated GPU causes the fans to go mental and the laptop battery life to last an hour max. :\

**Comment by Akemi** [comment]
yeah from this lldb output it looks like it can't create a pixel format successfully. i see that there is one small bug that is probably the reason of the `llegal instruction: 4` crash instead of gracefully exiting and returning an error message. [here](https://github.com/mpv-player/mpv/blob/master/video/out/cocoa-cb/video_layer.swift#L177) i should probably test for `pix == nil`. though that only as a side note.

i will have a look at it tomorrow, though if you feel adventurous you can try to figure out which of the new attributes is causing your problem. [new](https://github.com/mpv-player/mpv/blob/master/video/out/cocoa-cb/video_layer.swift#L157) pixel format and [old](https://github.com/mpv-player/mpv/blob/master/video/out/opengl/context_cocoa.c#L72) pixel format the only two differences are the `kCGLPFADoubleBuffer` and `kCGLPFABackingStore` attributes. try to remove both individually and if that doesn't help remove both together, the new swift file that is. let me know of your findings if you decide to try.

**Comment by aitte2** [comment]
I ran this to do each recompile: `rm -f build/osdep/macOS_swift.o && rm -f build/mpv && ./waf`, and I used `lldb -- ./build/mpv` to run the program.

With each of those lines commented out (both, or either), it still crashes with the exact same error.

This is beyond my skill, and I pray that this isn't some issue where Swift itself forces the dedicated GPU...

PS: A tech article which is possibly slightly related but of no use to 2010 MacBooks. https://developer.apple.com/library/content/qa/qa1734/_index.html

**Comment by AirPort** [comment]
>This is beyond my skill, and I pray that this isn't some issue where Swift itself forces the dedicated GPU...

I don't think so, not on a general basis at least: although admittedly I'm on a newer system version (10.13.3) and maybe slightly different hardware, with the precise same swift version I'm perfectly able to run mpv on the iGPU. I'd blame the OS more than anything else.

**Comment by aitte2** [comment]
@AirPort What hardware are you using?

I suspect that all MacBook Pro 2010 users are affected but not newer machines.

**Comment by aitte2** [comment]
If it helps, I tracked down the CPU which tells us what Intel GPU it is:

https://ark.intel.com/products/43560/Intel-Core-i7-620M-Processor-4M-Cache-2_66-GHz

![gpu](https://user-images.githubusercontent.com/18485256/37333710-512149de-26aa-11e8-89c0-d22f0dd9003a.png)

**Comment by aitte2** [comment]
PS @Akemi what you suggested earlier about "nil" is definitely a worthwhile change:

```diff
diff --git a/video/out/cocoa-cb/video_layer.swift b/video/out/cocoa-cb/video_layer.swift
index 8cdb692ae5..e00edcd58d 100644
--- a/video/out/cocoa-cb/video_layer.swift
+++ b/video/out/cocoa-cb/video_layer.swift
@@ -174,7 +174,7 @@ class VideoLayer: CAOpenGLLayer {
             }
         }
 
-        if err != kCGLNoError {
+        if pix == nil || err != kCGLNoError {
             fatalError("Couldn't create CGL pixel format: \(CGLErrorString(err)) (\(err))")
         }
         return pix!
```

Result:

```
./mpv
Fatal error: Couldn't create CGL pixel format: 0x00007fff864d51e1 (_CGLError(rawValue: 0)): file ./mpv-gputest/video/out/cocoa-cb/video_layer.swift, line 178
Illegal instruction: 4
```

We still see illegal instruction but at least it gives a meaningful error before that now.

**Comment by aitte2** [comment]
I noticed something else which is actually very serious:

This whole swift change breaks even on machines that don't use "Force Integrated GPU only" via gfxCardStatus.

Because after the swift change, mpv always tries to enumerate dedicated graphics cards at app startup, but the OS doesn't always switch cards fast enough.

So, a user opens mpv while their machine is still in Integrated mode, and then mpv queries dedicated graphics cards, which triggers the OS to make a switch to the dedicated GPU, but before the switch is completed mpv has already enumerated the cards and failed with "Illegal Instruction: 4".

So in its current state, the new release (latest commits after 0.28.2) will not work on MacBook Pro 2010s even on totally stock, unmodified machines. :\

It also has a sad side effect: Because mpv now runs that enumeration code at every startup, mpv always makes the OS auto-enable the dedicated GPU even when no video output is used. So for example running `mpv 'soundcloud link'` on the command line to play some audio will enable the dedicated GPU now...

Let's hope the swift code can be edited to make it work on the integrated GPU... Or perhaps the video layer can be moved back to the old OpenGL backend without any loss? The 230 lines of video_player.swift code could perhaps use the old backend instead, for most of the features.

I'll leave it at this for now and await your instructions, or if you want to chat today (I am only available today though).

**Comment by maru-sama** [comment]
I am wondering about the deprecation error. Is there a difference if you start it with  --vo=opengl-cb as in

./mpv --idle --force-window --vo=opengl-cb

**Comment by aitte2** [comment]
@maru-sama Yeah, I was using `vo=gpu`. With your command I got rid of the deprecation error.

I don't know the difference because my edits to the Swift video_player affected even `vo=gpu`...

Anyway, the deprecation message is gone but the intermittent inability to start up even in "dynamic GPU switching mode" still remains. And like vo=gpu, the vo=libmpv fails 100% of the time if "integrated GPU only" is used.

```
~/Downloads/mpv-gputest/build
+ $ ./mpv --idle --force-window --vo=libmpv


Exiting... (Quit)

~/Downloads/mpv-gputest/build
+ $ ./mpv --idle --force-window --vo=libmpv
Illegal instruction: 4
```

From now on I will do all testing with vo=libmpv though, even though it didn't help...

**Comment by AirPort** [comment]
Made a quick tests on an old 2010 MBP with High Sierra, I can confirm that forcing the iGPU with gfxcardstatus makes mpv master crash with the cocoa-cb backend, but with the stock configuration (dynamic switching) it works just fine.

**Comment by aitte2** [comment]
Thanks a lot for confirming the results! Excellent. Now we know that the new backend isn't working on more than just this machine...

> but with the stock configuration (dynamic switching) it works just fine

This is not exactly right. ;) Maybe 30 out of 31 launches are correct, but there is a race condition where it dies even in stock (dynamic) mode because the dedicated GPU doesn't spin up fast enough. I demonstrated it above. But it does take a lot of attempts (it is very intermittent... sometimes mpv crashes almost every launch, sometimes it can take minutes of relaunching over and over until I see it again...).

**Comment by aitte2** [comment]
Good news @Akemi this is a problem in mpv, not in Swift.

I have no idea how Xcode works so I struggled a lot, but I did this:

- Set gfxCardStatus to "Integrated GPU Only".
- `git clone https://github.com/joshb/CocoaOpenGL-Swift`
- Open it in Xcode and go to menu: View - Navigators - Show Project Navigator.
- Double-click the project name at the top of the list (this took me 15 minutes to find, sigh).
- Scroll down to Deployment Info and set Target to 10.12 (it was default to 10.13).
- Close the window.
- Go to menu: Product - Build For - Running.
- In the bottom of the project navigator will be a "Products" area which says "CocoaOpenGL-Swift-macOS.app". Right click it and "Show in Finder".
- Lastly, just double click it to run it...

Result: A rotating, colored Swift "Cocoa OpenGL" cylinder. Running on my Intel HD graphics.

This gives me hope that mpv has a small setting/bug that can be fixed.

**Comment by aitte2** [comment]
At akemi's request here's a ticket summary of the 3 GPU-related issues:

1. (Happening with automatic GPU switching): The race condition where 5% of the startups lead to a "Illegal Instruction: 4" crash because macOS seemingly does not enable the dedicate GPU fast enough.
2. (Happening with "Integrated GPU Only" mode): Mpv is unable to initialize an OpenGL surface. But thankfully it's doable (there's just some tiny bug in mpv), since I successfully ran another project's Swift OpenGL surface above on the integrated GPU.
3. (Happening with automatic GPU switching): The whole player triggers the dedicated GPU now even if just playing an audio file on the command line.

Akemi has fixed 1 and 2 (it was some bug). :D ❤️ 

About 3: It's because mpv pre-allocates an offscreen OpenGL context for later "already-prepared" usage in case video output is necessary. And Apple sees that as a reason to trigger 3D. Which on 2010 Mac laptops will enable the dedicated GPU. On 2011 or later laptops it will not trigger the dedicated GPU.

So in short: 1 and 2 are critical issues but are fixed (soon :D), and 3 is not critical since it only affects 2010 MacBook Pros, and the owners of those (such as me and Akemi) can just use "integrated mode only" via gfxCardStatus so that mpv never triggers the dedicated GPU.

Awesome. 👍

**Comment by aitte2** [comment]
Okay through some great teamwork @Akemi and I have fixed it.

https://github.com/Akemi/mpv/commit/2f81921685c1def456fec2ccc6e04f12178568bd (that still has some debug lines left)

The bug was that Apple doesn't return any pixel format on macOS 10.12 in "integrated gpu only" mode, but also doesn't return any error. So the error check in the loop never got triggered, and therefore a null (nil) format was incorrectly returned.

The fix was literally to add checking against `pix == nil`. With this fix, bug 2 is fixed.

I will investigate bug 1 but I suspect it's fixed too. I suspect that 1 was literally due to macOS _sometimes_ returning a nil pixel format with no error _even when automatic GPU switching was on_.

About 3: Akemi _may_ try to restructure things to see if it doesn't trigger the GPU that way. But either way, it's no big deal. Anyone with a MacBook Pro 2011 or newer will automatically always run mpv on the integrated GPU anyway, and a MacBook Pro 2010 can be forced to be on integrated GPU via gfxCardStatus so it's no big problem. And there aren't many of those machines in active use in the world anymore, so it's okay if that part is unfixable.
